File: LeetCode/Java/1160-Unknown-拼写单词.java

/*
 * @lc app=leetcode.cn id=1160 lang=java
 *
 * [1160] 拼写单词
 */
import java.util.*;

// @lc code=start
class Solution {
    public int countCharacters(String[] words, String chars) {
        int res = 0;

        char[] cs = new char[26];
        for (char c : chars.toCharArray())
            ++cs[c - 'a'];

        for (String s : words) {
            boolean flag = true;
            if (chars.length() < s.length())
                flag = false;
            else {
                char[] t = new char[26];
                for (char c : s.toCharArray()) {
                    ++t[c - 'a'];
                    if (t[c - 'a'] > cs[c - 'a']) {
                        flag = false;
                        break;
                    }
                }
            }
            if (flag)
                res += s.length();
        }

        return res;
    }
}
// @lc code=end

File: LeetCode/C/1160-Unknown-拼写单词.cpp
/*
 * @lc app=leetcode.cn id=1160 lang=cpp
 *
 * [1160] 拼写单词
 */
#include <string>
#include <vector>
using namespace std;

// @lc code=start
class Solution {
  public:
    int countCharacters(vector<string> &words, string chars) {
        int ch[26] = {0};
        for (char c : chars)
            ++ch[c - 'a'];
        int res = 0;
        for (string s : words) {
            bool match = true;
            int count[26] = {0};
            for (char c : s)
                if (++count[c - 'a'] > ch[c - 'a']) {
                    match = false;
                    break;
                }
            if (match)
                res += s.length();
        }
        return res;
    }
};
// @lc code=end
